odoo/addons/base/models/ir_mail_server.py

```diff
@@ -653,6 +653,10 @@ def _find_mail_server(self, email_from, mail_servers=None):
         if mail_server:
             return mail_server[0], email_from
 
+        mail_server = mail_servers.filtered(lambda m: email_domain_extract(m.from_filter) == email_from_domain)
+        if mail_server:
+            return mail_server[0], mail_server[0].from_filter
+
```

Comment:
this isn't how it's suppose to work, it that case it should take the notification email server

```
Imagine I have 2 mail servers:

| from_filter      | priority |
|------------------|----------|
| user@domain1.com | 10       |
| user@domain2.com | 20       |

Now, if I send an email from `other_user@domain2.com`, what server do you think it should be selected?

Before this patch, the 1st server would be selected because there's no exact match with `other_user@domain2.com`.

After this patch, the 2nd mail server will be selected if possible, because it's a closer fallback than the 1st one.
```

in your commit message, the configuration is wrong, the from filter should be `domain2.com`


the notification email should be used in that case and it's not a bug

Reply:
Hello! I see that either you didn't understand the issue or I don't understand you. Let me explain a bit better.

My need is pretty simple: everything ending in domain1.com needs to be rewritten to user@domain1.com, and everything ending in domain2.com needs to be rewritten to user@domain2.com.

Such a simple configuration is impossible to do without this PR. If it's a configuration problem, then please tell me which configuration I need to use.

The reason why this is needed is that the companies' policy forbids mail spoofing, and it's a multi-company DB where each company has a different domain.

Reply:
@yajo But how your change will encapsulate the email with both emails address ?

The code that does that is
```
smtp_from = encapsulate_email(message['From'], notifications_email)
```


And you can only encapsulate into the notification email


Your fix isn't encapsulating the email, it's changing the mail server. So it won't work properly.

To have that behavior, we would need to add a new field on the outgoing mail server, "Encapsulation Email" or something like that that take the priority on `_get_default_from_address`, but your current code won't work anyway, and will even break existing flow.

Reply:
It's not necessary to encapsulate. As you can see in this PR, I'm falling back to using the from filter address' domain to match a *closer* fallback smtp server. Then I'm returning the mail server's full `from_filter` as the from address.

This way, for example, if a mail server's from filter is `user@domain2.com` and the original from address is `bob@domain2.com`, that mail server will be used as a *closer fallback* because domains match. But then, the from address will be forced to `user@domain2.com`.

That value returned by this method is obtained when connecting to the smtp server:
https://github.com/odoo/odoo/blob/91f213e202ca29c4ef0bb190663fc48f169a64a6/odoo/addons/base/models/ir_mail_server.py#L307

Then it's saved in the connection object as an attribute:
https://github.com/odoo/odoo/blob/91f213e202ca29c4ef0bb190663fc48f169a64a6/odoo/addons/base/models/ir_mail_server.py#L409

Finally, it's used as the smtp from address. Encapsulation only happens if it matches the notifications email, which might or not happen:
https://github.com/odoo/odoo/blob/91f213e202ca29c4ef0bb190663fc48f169a64a6/odoo/addons/base/models/ir_mail_server.py#L593-L599

Just test the patch! It works 😃

Reply:
but the from_filter field purpose is not to be used as email from

as I said I would prefer to remove this condition, and to add a field on the outgoing mail server that overwrite the notification email, so the behavior will be explicit

I don't like this fix because it break existing flow. A mail server will take the priority over the default notification server if the domain match...

If you have 2 mail servers
- alice@gmail.com
- notification@mycompany.com => default notification server

And I send a email with "bob@gmail.com", it will change the email to "alice@gmail.com" right ?

Reply:
but I agree it would be cool tu spoort this feature :) (I just disagree on the implementation)

Reply:
Nice, I think at least both of us are now understanding each other! 😄 

> but the from_filter field purpose is not to be used as email from

In practical terms, when the from_filter is an address, that's what it does now. Actually in https://github.com/odoo/odoo/pull/104417 you see another case where this logic also needed to be applied.

However I agree that being explicit is always better. That way, there'd be one field for matching the mail server and another one for setting the from address.

Do you want me to open that PR for master, or do you prefer to do it yourself?

Reply:
#104417 this is only to test the connection (so it's okish even if not perfect)

> However I agree that being explicit is always better. That way, there'd be one field for matching the mail server and another one for setting the from address.

yes exactly :) 

I'll do it so it follow our testing / review work flow :)

Here is the PR if you want to follow it: https://github.com/odoo/odoo/pull/109629

(I'll work on it when I'll have some time)

Thanks again for your idea :)

Reply:
Thank you very much 😊

Reply:
@yajo Re :) 

In the end it will be handle with: https://github.com/odoo/odoo/pull/76734

With a company_id field on the mail servers, so your use case should be possible with this :) 

(the idea of a custom mail.default.from per mail servers would allow even more customization, but we think the email configuration is already complicate enough)

Cheers